aioxmpp/stream.py

```diff
@@ -927,31 +927,46 @@ def _destroy_stream_state(self, exc):
             self.on_stream_destroyed(exc)
             self._established = False
 
-    def _iq_request_coro_done(self, request, task):
+    def _send_iq_reply(self, request, result):
+        if isinstance(result, errors.XMPPError):
+            response = request.make_reply(type_=structs.IQType.ERROR)
+            response.error = stanza.Error.from_exception(result)
+        else:
+            response = request.make_reply(type_=structs.IQType.RESULT)
+            response.payload = result
+        self._enqueue(response)
+
+    def _iq_request_coro_done_remove_task(self, task):
+        self._iq_request_tasks.remove(task)
+
+    def _iq_request_coro_done_send_reply(self, request, task):
         """
         Called when an IQ request handler coroutine returns. `request` holds
         the IQ request which triggered the excecution of the coroutine and
         `task` is the :class:`asyncio.Task` which tracks the running coroutine.
 
         Compose a response and send that response.
         """
-        self._iq_request_tasks.remove(task)
         try:
             payload = task.result()
         except errors.XMPPError as err:
-            response = request.make_reply(type_=structs.IQType.ERROR)
-            response.error = stanza.Error.from_exception(err)
+            self._send_iq_reply(request, err)
         except Exception:
             response = request.make_reply(type_=structs.IQType.ERROR)
             response.error = stanza.Error(
                 condition=errors.ErrorCondition.UNDEFINED_CONDITION,
                 type_=structs.ErrorType.CANCEL,
             )
+            self._enqueue(response)
```

Comment:
How about re-using ``_send_iq_reply`` here (with the created ``stanza.Error`` object)?

Reply:
Then I would have to extend the de-multiplexing to make an error reply for certain payload XSOs passed in (but this should be possible via XSO inspection: if it is registered for error or for payload).

Reply:
aahh, right, yeah, keep it this way, sorry for the Confusion.